[rocketninja7] iP

rocketninja7 opened the pull request:
# ip bot
A simple bot to **keep track of your tasks** for you!

> "The best thing is not to fight at all" - Ip Man

To begin:
1. Download the [jar](https://github.com/rocketninja7/ip/releases) file.
2. Run ```java -jar //path/to/jar```
3. Enjoy! 👍 

Types of tasks supported:
- ToDo
- Deadline
- Event

Features:
- [x] Add tasks
- [x] Delete tasks
- [x] Mark tasks
- [x] Unmark tasks
- [x] Find tasks
- [ ] Edit tasks
- [ ] ~Martial arts~

For developers, maybe you can start at the main method of Duke.java:
```java
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

rocketninja7 committed:
Changed name of chatbot and implemented intial skeleton version

rocketninja7 committed:
Improved bot to echo commands entered by user

rocketninja7 committed:
Added ability to store text and display them

rocketninja7 committed:
Added Task object and ability to mark and unmark tasks

rocketninja7 committed:
Added support for ToDos, Events and Deadlines

rocketninja7 committed:
Added semi automated testing using I/O redirection

rocketninja7 committed:
Added error handling with Exceptions

rocketninja7 committed:
Added support for deleting tasks

rocketninja7 committed:
Refactored code using enums

rocketninja7 committed:
Add save functionality

rocketninja7 committed:
Merge branch 'branch-Level-7'
Merge save feature to master

rocketninja7 committed:
Use dates/times instead of String

rocketninja7 committed:
Merge branch 'branch-Level-8'
Merge date and time updates into master

rocketninja7 committed:
Added command to print deadlines/events on certain date

rocketninja7 committed:
Merge branch 'branch-Level-8'
Merge command to print deadlines/events on specific date to master

rocketninja7 committed:
Make the code more OOP

rocketninja7 committed:
Improve OOP

rocketninja7 committed:
Merge branch 'develop'
Merge OOP updates to master

rocketninja7 committed:
Divided classes into packages

rocketninja7 committed:
Merge branch 'develop'
Merge added packages to master

rocketninja7 committed:
Merge branch 'add-gradle-support' into develop
Add gradle support

rocketninja7 committed:
Merge branch 'develop'
Merge Gradle to master

juliusgambe commented on src/main/java/ipbot/Duke.java:
```diff
@@ -0,0 +1,149 @@
+        this.ui = new Ui();
+    }
+
+    public void run() {
```
You may want to add a javadoc here :D

juliusgambe commented on src/main/java/ipbot/util/Storage.java:
```diff
@@ -0,0 +1,66 @@
+import java.io.IOException;
+import java.time.format.DateTimeParseException;
+
+public class Storage {
```
Please add javadocs for your public classes and methods :D

juliusgambe commented on src/main/java/ipbot/util/TaskList.java:
```diff
@@ -0,0 +1,122 @@
+        return deadline;
+    }
+
+    public Event addEventWithArgs(Map<String, String> commandArgs) throws CommandArgumentException {
```
Perhaps a better name is `addEventWithCommands`

juliusgambe commented on src/main/java/ipbot/Duke.java:
```diff
@@ -0,0 +1,149 @@
+                            case TODO:
+                                if (commandArgs.isEmpty()) {
+                                    throw new CommandArgumentException("Task description cannot be empty!");
+                                }
+                                ToDo toDo = taskList.addToDoWithArgs(commandArgs);
+                                ui.printAddedItem(toDo, "todo");
+                                break;
+                            case DEADLINE:
+                                Deadline deadline = taskList.addDeadlineWithArgs(commandArgs);
+                                ui.printAddedItem(deadline, "deadline");
+                                break;
+                            case EVENT:
+                                Event event = taskList.addEventWithArgs(commandArgs);
+                                ui.printAddedItem(event, "event");
+                                break;
+                            case DELETE:
+                                int deleteIndex = Parser.checkIndexArg(commandArgs.get(""), taskList.getTasksSize());
+                                if (deleteIndex == -1) {
+                                    throw new CommandArgumentException("Invalid task to delete: " + commandArgs);
+                                }
+                                Task task = taskList.deleteTask(deleteIndex);
+                                ui.printDeletedItem(task);
+                                break;
+                        }
+                    }
+                }
+            }
+            catch(CommandArgumentException e){
+                System.out.println(e.getMessage());
+            }
+            catch (DateTimeParseException e) {
+                System.out.println(e.getMessage());
+            }
+            catch (IllegalArgumentException e) {
+                System.out.println(e.getMessage());
+            }
+            ui.printDivider();
+            this.storage.writeTasksToFile(taskList);
+        }
+    }
```
You may want to abstract different parts of this code to a Parser :)

juliusgambe commented on src/main/java/ipbot/util/Parser.java:
```diff
@@ -0,0 +1,42 @@
+                argEndIndex = currArg.length();
+            }
+            String argName = currArg.substring(0, argEndIndex).toLowerCase();
+            String argArgs = currArg.substring(argEndIndex).strip();
```
You may want to provide a more specific name for `argArgs`

juliusgambe reviewed (commented):
Generally, I think there can be improvements made to code cleanliness in terms of abstracting different functionalities in the code. Furthermore, I think you may need to add javadocs more to explain your classes and methods :) Overall, I think it is interesting you incorporated a lot of other data structures into the project and explored more than what is required of the module! I wish I had as much time as you haha :D

rocketninja7 committed:
Added JUnit tests for Ui and Parser

rocketninja7 committed:
Merge branch 'develop'
Merge JUnit tests into master

rocketninja7 committed:
Changed tests and gradle to make jar work

rocketninja7 committed:
Merge branch 'develop'
Merge gradle and test edits to make jar work into master

rocketninja7 committed:
Added JavaDocs

rocketninja7 committed:
Conform to coding standards

rocketninja7 committed:
Add find feature

rocketninja7 committed:
Merge branch 'branch-Level-9'
Merge find feature to master

rocketninja7 committed:
Merge branch 'branch-A-JavaDoc'

rocketninja7 committed:
Merge branch 'branch-A-CodingStandard'

butteredyakiimo commented on src/main/java/ipbot/Duke.java:
```diff
@@ -0,0 +1,154 @@
+                } else {
+                    String command = commandResult.getFirst();
+                    Command commandName = Command.commandEnum(command);
+                    if(commandName == null){
```
Remember to add a white space after Java reserved keywords!:D

butteredyakiimo commented on src/main/java/ipbot/model/Command.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Represents a command available in ipbot.
+ */
+public enum Command {
```
I like how you made use of enums! Made the switch statement in Duke read nicely

butteredyakiimo reviewed (commented):
LGTM :+1: OOP was done well, and code quality is good. There are no big violations of the coding standard, but as mentioned by the other reviewer, remember to add java docs for all public methods and classes!

rocketninja7 committed:
Add JavaFX GUI to the chatbot

rocketninja7 committed:
Remove dead code from Ui class

rocketninja7 committed:
Add Launcher class so that javafx modules do not have to be modules

rocketninja7 committed:
Merge GUI to master

rocketninja7 committed:
Improve code quality

Some methods are too long due to lack of SLAP. Some
magic Strings are also used.

Using SLAP allows us to improve readability of code.
Using constants also makes code easier to maintain.

Let's pull up all error checking code into separate
error checking methods. Let's also change all magic
Strings to constants.

Putting error checking code into separate methods
allows the method to not be cluttered with code
that does not perform its main function.

rocketninja7 committed:
Add assertions to code

Code does not have assertions.

Certain methods can be called with incorrect values,
resulting in unhandled errors that cause undefined
behaviour in the program.

Let's add assertions to the code.

Using assertions allows us to give a readable error
message for debugging.

rocketninja7 committed:
Merge pull request #2 from rocketninja7/branch-A-CodeQuality

Improve code quality

rocketninja7 committed:
Merge branch 'branch-A-Assertions'

rocketninja7 committed:
Add feature to recur tasks

rocketninja7 committed:
Merge branch 'branch-B-RecurringTasks'

rocketninja7 committed:
Resolve TODOs

rocketninja7 committed:
Add Ui.png

rocketninja7 committed:
Update User Guide

rocketninja7 committed:
Refactor code and fix bugs

rocketninja7 committed:
Update JavaDocs

rocketninja7 committed:
Update gradle config to make app run on Linux
